### PR TITLE
Fixed #30736 -- Added Storage.get_alternative_name() to allow customization.

### PR DESCRIPTION
**Pull request by yukihira1992**
… resolving algorithm.

ticket [#30736](https://code.djangoproject.com/ticket/30736)

### CODE REVIEW

File: django/core/files/storage.py [hunk 1]
```diff
@@ -60,21 +60,28 @@ def get_valid_name(self, name):
...
         Return a filename that's free on the target storage system and
         available for new content to be written to.
         """
         dir_name, file_name = os.path.split(name)
         file_root, file_ext = os.path.splitext(file_name)
-        # If the filename already exists, add an underscore and a random 7
-        # character alphanumeric string (before the file extension, if one
-        # exists) to the filename until the generated filename doesn't exist.
+        # If the filename already exists, generate an alternative filename
+        # by get_alternative_name() until the generated filename doesn't exist.
```
Comment (felixxm) [inline django/core/files/storage.py]:
```
# If the filename already exists, generate an alternative filename
# until it doesn't exist.
```

File: django/core/files/storage.py [hunk 2]
```diff
@@ -60,21 +60,28 @@ def get_valid_name(self, name):
         """
         return get_valid_filename(name)
 
+    def get_alternative_name(self, file_root, file_ext):
+        """
+        Return an alternative filename candidate, adding an underscore and
+        a random 7 character alphanumeric string (before the file extension,
+        if one exists) to the filename.
```
Comment (felixxm) [inline django/core/files/storage.py]:
```
Return an alternative filename, by adding an underscore and a random 7
character alphanumeric string (before the file extension, if one
exists) to the filename.
```

**Comment by yukihira1992** [comment]
@felixxm Thanks, I updated the doc.

**Comment by felixxm** [comment]
@yukihira1992 Thanks :+1: I added release notes and updated `howto/custom-file-storage.txt`. Welcome aboard :boat:

**Pull request merged**